tools/debug_tool/static/pipeline_viewer.js

```diff
@@ -3,27 +3,63 @@ function isIE() { return ((navigator.appName == 'Microsoft Internet Explorer') |
 // Time between updating kotekan metrics
 const POLL_WAIT_TIME_MS = 1000;
 // Poll kotekan via web server every POLL_WAIT_TIME_MS and update page with new metrics
-function poll(label, index) {
-    get_data().then(function (new_buffers) {
-        update_buf_utl(new_buffers, label, index);
+function poll(buffer_labels, stage_labels) {
+    get_data("/buffers").then(function (new_buffers) {
+        update_buf_utl(new_buffers, buffer_labels);
     });
 
+    get_data("/cpu_ult").then(function (cpu_stats) {
+        update_cpu_utl(cpu_stats, stage_labels);
+    })
+
     // Call poll() again to get the next message
-    setTimeout(() => { poll(label, index); }, POLL_WAIT_TIME_MS);
+    setTimeout(() => { poll(buffer_labels, stage_labels); }, POLL_WAIT_TIME_MS);
 }
 
 // Update buffer utilization
-function update_buf_utl(new_buffers, label, index){
-    for (var i of index) {
-        var name = d3.select(label[0][i]).select("tspan").text();
-        d3.select(label[0][i]).select("#utl")
+function update_buf_utl(new_buffers, label){
+    label[0].reduce((pre, cur) => {
+        var el = d3.select(cur);
+        var name = el.select("tspan").text();
+        el.select("#utl")
             .text(new_buffers[name].num_full_frame + "/" + new_buffers[name].num_frames);
+    })
+}
+
+// Update stage utilization
+function update_cpu_utl(cpu_stats, label){
+    if ("error" in cpu_stats) {
+        return;
     }
+
+    label[0].reduce((pre, cur) => {
+        var el = d3.select(cur);
+        var name = el.select("tspan").text();
+        var tot_ult = 0, usr_ult = 0, sys_ult = 0;
+
+        d3.select(cpu_stats[name]).forEach((stage) => {
+            var threads = Object.keys(stage[0]);
+            for (tid of threads) {
+                var usr = stage[0][tid]["usr_cpu_ult"];
+                var sys = stage[0][tid]["sys_cpu_ult"];
+                usr_ult += usr;
+                sys_ult += sys;
+                tot_ult += usr + sys;
+            }
```

Comment:
For now this works fine.  However, if we have time it might be nice to break out each thread for multi-threaded stages. And also indicate the which CPUs the threads are allowed to run on (e.g. thread CPU affinity), however that requires more changes on the C++ side, so I think that can be another PR later when we have time.

Reply:
That's a good idea!